crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -876,13 +854,161 @@ impl Ics2ClientState for ClientState {
+    /// [this](https://ibc.cosmos.network/main/ibc/upgrades/quick-guide.html)
+    /// guide
+    fn verify_upgrade_client(
         &self,
-        _consensus_state: Any,
-        _proof_upgrade_client: RawMerkleProof,
-        _proof_upgrade_consensus_state: RawMerkleProof,
+        upgraded_client_state: Any,
+        upgraded_consensus_state: Any,
+        proof_upgrade_client: RawMerkleProof,
+        proof_upgrade_consensus_state: RawMerkleProof,
+        root: &CommitmentRoot,
+    ) -> Result<(), ClientError> {
+        // Make sure that the client type is of Tendermint type `ClientState`
+        let mut upgraded_tm_client_state = TmClientState::try_from(upgraded_client_state)?;
+
+        // Make sure that the consensus type is of Tendermint type `ConsensusState`
+        let upgraded_tm_cons_state = TmConsensusState::try_from(upgraded_consensus_state)?;
+
+        // Note: verification of proofs that unmarshalled correctly has been done
+        // while decoding the proto message into a `MsgEnvelope` domain type
+        let merkle_proof_upgrade_client = MerkleProof::from(proof_upgrade_client);
+        let merkle_proof_upgrade_cons_state = MerkleProof::from(proof_upgrade_consensus_state);
+
+        // Make sure the latest height of the current client is not greater then
+        // the upgrade height This condition checks both the revision number and
+        // the height
+        if self.latest_height() >= upgraded_tm_client_state.latest_height() {
+            return Err(ClientError::LowUpgradeHeight {
+                upgraded_height: self.latest_height(),
+                client_height: upgraded_tm_client_state.latest_height(),
+            });
+        }
+
+        // Ensure that the new unbonding period is not shorter than the current's
+        // client unbonding period
+        if self.unbonding_period > upgraded_tm_client_state.unbonding_period {
+            return Err(ClientError::ClientSpecific {
+                description:
+                    "cannot upgrade client with a shorter unbonding period than the current one"
+                        .to_string(),
+            });
+        }
+
+        // Check to see if the upgrade path is set
+        let mut upgrade_path = upgraded_tm_client_state.clone().upgrade_path;
+        if upgrade_path.pop().is_none() {
+            return Err(ClientError::ClientSpecific {
+                description: "cannot upgrade client as no upgrade path has been set".to_string(),
+            });
+        };
+
+        let last_height = self.latest_height().revision_height();
+
+        // Construct the merkle path for the client state
+        let mut client_upgrade_path = upgrade_path.clone();
+        client_upgrade_path.push(ClientUpgradePath::UpgradedClientState(last_height).to_string());
+
+        let client_upgrade_merkle_path = MerklePath {
+            key_path: client_upgrade_path,
+        };
+
+        upgraded_tm_client_state.zero_custom_fields();
```

Comment:
why are we zeroing some fields here? [ibc-go doesn't seem](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/upgrade.go#L75-L84) to be doing that

Reply:
When a chain creates an `UpgradedClientState` through an upgrade proposal, it should zero out all the cleint-customizable fields ([Ref](https://github.com/cosmos/ibc-go/blob/main/docs/ibc/upgrades/quick-guide.md#step-by-step-upgrade-process-for-sdk-chains)). The upgrade client proof is for that zeroed-out client state, and to validate against, we also should zero out respective fields of `UpgradedClientState`. 
Check [here](https://github.com/cosmos/ibc-go/blob/e093d85b533ab3572b32a7de60b88a0816bed4af/modules/core/keeper/msg_server_test.go#L678) please

Reply:
Interesting, your references point to the `main` branch, [which changed since v5.0.1](https://github.com/cosmos/ibc-go/blob/c0acd5bd1778f2b7ecdf593006f56bd3e273bd49/modules/light-clients/07-tendermint/types/upgrade.go#L76), the version we are targeting compatibility with (notice how they don't zero the fields there). I'm not sure which version we should target anymore for ibc-rs chains to be compatible with ibc-go chains (as it depends on the version of ibc-go that the chain is using).

Reply:
#391